src/caffeine/allocation_m.f90

```diff
@@ -2,13 +2,33 @@
 ! Terms of use are as specified in LICENSE.txt
 module allocation_m
     use iso_c_binding, only: c_ptr, c_int, c_intmax_t, c_size_t, c_funptr
+
     implicit none
     private
-    public :: prif_allocate, prif_allocate_non_symmetric, prif_deallocate, prif_deallocate_non_symmetric
+    public :: &
+        prif_coarray_handle, &
+        prif_allocate, &
+        prif_allocate_non_symmetric, &
+        prif_deallocate, &
+        prif_deallocate_non_symmetric
 
+    ! TODO: Should these actually be interleaved if we use explicit size?
+    type, bind(C) :: cobound_pair
+      integer(c_intmax_t) :: lcobound, ucobound
+    end type
+    
+    type, bind(C) :: handle_data
+      type(c_ptr) :: coarray_data
+      integer(c_int) :: corank
+      integer(c_size_t) :: coarray_size
+      type(c_funptr) :: final_func
+      type(c_ptr) :: previous_handle, next_handle
+      type(cobound_pair) :: cobounds(15)
```

Comment:
TODO: Missing `type(c_ptr) :: context_data`

Probably also missing a pointer to the establishment team, once that type is defined

Otherwise looks like a good start!

Reply:
Actually, I don't think a coarray does need a pointer to the establishing team. All coarray accesses are either with respect to the current team (which is easily grabbed), or with respect to an explicitly provided team.

Reply:
Are you sure? What about a coarray established in a parent team and not reassociated in the CHANGE TEAM statement? Aren't coindexed expressions on such an array (with no explicit team) interpreted wrt the establishing team?

Reply:
> If a TEAM= specifier appears in an image-selector ...  If a TEAM_NUMBER= specifier appears in an image-selector ... Otherwise, the team of the image selector is the current team.

> An image selector shall specify an image index value that is not greater than the number of images in the team of the image selector, and identifies the image with that index in that team.

Reply:
>  Otherwise, the team of the image selector is the current team.

I don't understand how the coindexes are interpreted wrt to a current team that is a subset of the establishing team when corank > 1. However I assume this is explained somewhere..